### PR TITLE
feat: holding down CMD/CTRL will disable snap to grid when grid is active

### PR DESCRIPTION
**Pull request by zsviczian**
A possible partial solution to #6889 and a full solution to #4945

### CODE REVIEW

File: src/locales/en.json [hunk 1]
```diff
@@ -78,7 +78,7 @@
     "group": "Group selection",
     "ungroup": "Ungroup selection",
     "collaborators": "Collaborators",
-    "showGrid": "Show grid",
+    "showGrid": "Show grid (Cmd/Ctrl: No Snap)",
```
Comment (dwelle) [inline src/locales/en.json:81]: this should instead go to the toolbar hint

File: src/math.ts [hunk 2]
```diff
@@ -266,8 +266,9 @@ export const getGridPoint = (
   x: number,
   y: number,
   gridSize: number | null,
+  snapToGrid: boolean = false,
```
Comment (dwelle) [inline src/math.ts:269]: we can reuse the `gridSize` for this, by assigning `null` if we don't want to snap to grid. Not the most intuitive parameter (as it's currently being used already), but your proposal makes the grid snapping dependent on two separate parameters so it's making it even less clear.

**Comment by dwelle** [comment]
Hey! 

I wonder how it should play with https://github.com/excalidraw/excalidraw/pull/6256.

- when grid enabled, snapping disabled, that's easy -> no conflict, use snaps
- when grid enabled, snapping enabled, there's a conflict -> with `ctrl/cmd` the user will either want to disable snapping or grid or both, but we'll have choose just one. Currently (in the snapping PR), the behavior is that snapping is preferred over grid, so when you disable snapping via `ctrl/cmd`, you go back to snapping to grid.

File: src/locales/en.json (hunk 1 above)
Comment (zsviczian) [reply src/locales/en.json:81]: Do you mean add it here on condition gridSize !== null? What happens when multiple hints are to be displayed?

https://github.com/excalidraw/excalidraw/blob/6abf1942814ec002f916b8a17c52dd124531bc42/src/components/HintViewer.tsx#L21-L111

File: src/math.ts (hunk 2 above)
Comment (zsviczian) [reply src/math.ts]: I was contemplating this, but opted for making it explicit with snapToGrid. Now I changed it to null if ctrl_or_cmd is down.

**Comment by zsviczian** [comment]
> when grid enabled, snapping enabled, there's a conflict -> with ctrl/cmd the user will either want to disable snapping or grid or both, but we'll have choose just one. Currently (in the snapping PR), the behavior is that snapping is preferred over grid, so when you disable snapping via ctrl/cmd, you go back to snapping to grid.

I'd vote for disabling both types of snapping when ctrl/cmd is down.

**Comment by dwelle** [comment]
> I'd vote for disabling both types of snapping when ctrl/cmd is down.

Ok.

Actually, my first point is not as clear cut either:

> when grid enabled, snapping disabled, that's easy -> no conflict, use snaps

If you're in grid mode without snapping enabled and you press cmd/ctrl, user may either wanna disable grid-snapping or enable object-snapping, and it's not clear what, either.

**Comment by zsviczian** [comment]
> If you're in grid mode without snapping enabled and you press cmd/ctrl, user may either wanna disable grid-snapping or enable object-snapping, and it's not clear what, either.

Else we can put snapping in the appstate changing `gridSize: number | null;` to `grid: {size: number, snaps: boolean} | null` and add a shortcut to toggle snapping behavior.

**Comment by dwelle** [comment]
Pushed [`efdd0b2` (#6983)](https://github.com/excalidraw/excalidraw/pull/6983/commits/efdd0b29543f9ac6ebd14a174beb22f21c2edff7) for the hints.

The conditionals aren't quite correct, but sufficient for now. Proper implem will need to wait for https://github.com/excalidraw/excalidraw/pull/7007

---

We should be good to merge.

**Review by dwelle — APPROVED** [review]

**Comment by dwelle** [comment]
Merging, thanks @zsviczian!

**Pull request merged**